rubix-spi/src/main/java/com/qubole/rubix/spi/SyncClusterManager.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2019. Qubole Inc
+ * Licensed under the Apache License, Version 2.0 (the License);
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+package com.qubole.rubix.spi;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.Set;
+
+public abstract class SyncClusterManager extends ClusterManager
+{
+  private static Log log = LogFactory.getLog(SyncClusterManager.class);
+
+  private Set<String> currentNodes;
+
+  protected abstract boolean hasStateChanged();
```

Comment:
Current model will spike up CPU usage a lot. The ClusterManager methods are very frequently called, e.g. `getCurrentNodeName` will be called for every MB of data being read via the worker node. This converts to 1k calls per GB to `hasStateChanged` which does comparisons internally.

I think the optimal way would be work with Presto team to add the logic in Presto's NodeManager to return info about state change e.g. it could store a version which increments every time there is a change in membership. Based on that version, we can hasStateChanged value here.

Reply:
I've confirmed that the hasStateChanged() method returns on an == check of the existing and new worker nodes. There's no extra overhead for this. The only time extra work is done is when the worker nodes change.

Reply:
Wouldn't the check be on `ImmutableSet.equals` which is not a simple `==` check?

Reply:
The == check is the first line of the `ImmutableSet.equals` implementation.

Reply:
Ok, so you confirmed that over a period of time if cluster state does not change then Presto's NodeManager returns same object? That is a dependency on undocumented behaviour of NodeManager, we will need to way/test to catch if this behaviour changes in future otherwise slowness will creep in at runtime.

Reply:
Looking at https://github.com/prestosql/presto/blob/c4ef3a95309686d778c9f319d7e91be8b6a6faaa/presto-main/src/main/java/io/prestosql/metadata/DiscoveryNodeManager.java#L144, the Set object would change every 5sec

Reply:
Yes, every 5 seconds, the `ImmutableSet.equals` would do an equals check on its members and again return true (unless workers were added or removed) and further processing would be stopped. The equals check on the Node is a string compare on the node identifier, so not an expensive check.

Reply:
> Ok, so you confirmed that over a period of time if cluster state does not change then Presto's NodeManager returns same object?
Yes, I confirmed this with logging that I've since removed. Let me see if I can add a unit test that confirms too.

Reply:
Ok, every 5sec there will be extra work to compare all the elements of the Set. I still think that doing it in conjunction with Presto NodeManager, something like a callback that NodeManager calls when it sees change in membership would work out best.
The problem with current approach, multiple comparisons every 5sec, would amplify if we remove the synchronisation you have added for all methods of SyncClusterManager. Given that there will be 100s of parallel requests to these methods (as many as task runner threads running in Presto), current code will serialise them which will definitely have perf impact. And as soon as you remove sync, you will have the problem of these 100s of threads (potentially all but may not be always all) doing the extra comparisons every 5sec. So a solution that is deeply integrated with Presto would be better.

Tagging @sopel39 @losipiuk to get inputs if it would be acceptable to add `stateChangeListener` to `InternalNodeManager` in Presto.

Reply:
With the enormity of everything  rubix is doing, there’s no way comparing 100 strings every five seconds would add any measurable overhead. How about running it with presto to measure it?

Reply:
> I think the optimal way would be work with Presto team to add the logic in Presto's NodeManager to return info about state change

This just shifts the problem to Presto. Workers list is not a real-time list and requires some time to refresh (I think discovery requests are sent every 1s, but there is also blacklist of failed nodes refreshed independently).

On a general note, I'm not sure what problem are we trying to fix? Is that an issue that we cache workers list for 5s? Workers should join/depart cluster much less frequently. Cluster with nodes changing every 5s won't be stable anyway.

Even if the list is not current, queries should not fail.

Reply:
@sopel39 - this issue (https://github.com/qubole/rubix/issues/462) is a secondary issue in that it causes more queries to fail than otherwise would with caching enabled when a node becomes unresponsive and/or is subsequently restarted/terminated. The reason is that rubix is caching the set of worker nodes and using this to determine where to send cache read requests. Because the set is cached by rubix instead of always getting it from the NodeManager, more cache read requests will be sent to nodes that are down which leads to more query failures with an internal error reason. We've observed this behavior consistently on a heavily loaded test cluster, while without caching enabled under heavy load, queries instead fail with an insufficient resources timeout. Preventing the proliferation of these internal error type failures is important because otherwise engineers that are on call start to get paged.

This PR solves this by always getting the list of worker nodes from the NodeManager and only doing the additional computation needed by rubix when the set of worker nodes changes. @stagraqubole is worried that this will cause too much overhead and is asking if it's feasible to have some kind of callback in NodeManager when the set of active workers changes. Looking closer at DiscoveryNodeManager, though it [polls for the node state every 5 seconds](https://github.com/prestosql/presto/blob/c4ef3a95309686d778c9f319d7e91be8b6a6faaa/presto-main/src/main/java/io/prestosql/metadata/DiscoveryNodeManager.java#L144), it seems that it only updates the set of nodes [if they've changed](https://github.com/prestosql/presto/blob/c4ef3a95309686d778c9f319d7e91be8b6a6faaa/presto-main/src/main/java/io/prestosql/metadata/DiscoveryNodeManager.java#L263). Irregardless of this, even if it did update every 5 seconds, I don't think the string compare of an equals operation occurring every 5 seconds would add any overhead.

The primary issue (which is being worked on by @stagraqubole) is [fixing the disk usage accounting](https://github.com/qubole/rubix/pull/464) to prevent out of disk space issues.

Other primary issues are:
- [Preventing worker nodes from being overwhelmed](https://github.com/qubole/rubix/issues/469) by dynamically sizing the thread pools. We've worked around this by configuring the thread pools down.
- [Removing cache files](https://github.com/qubole/rubix/issues/454) prior to calculating allowed disk usage. We've worked around this by changing our presto startup script.